src/python/WMCore/MicroService/MSPileup/MSPileupData.py

```diff
@@ -0,0 +1,177 @@
+    "fullReplicas": integer,  # total number of replicas to keep on Disk
+    "active": boolean,
+    "pileupSizeGB": integer, current size of the pileup in GB (integer)
+    "rulesList: list of strings (rules) used to lock the pileup id
+}
+
+The data flow should be done via list of objects, e.g.
+[<pileup object 1>, <pileup object 2>, ..., <pileup object n>]
+"""
+
+# third party modules
+from pymongo import IndexModel, errors
+
+# WMCore modules
+from WMCore.Database.MongoDB import MongoDB
+from WMCore.MicroService.MSCore.MSCore import MSCore
+
+
+def stripKeys(docs, stripKeys=['_id']):
+    """
+    Helper function to strip out keys from given dictionary. The input parameter
+    is a dictionary and keys are wiped out in place
+    :param docs: either input dictionary or list of dictionaries
+    :param stripKeys: list of of strip keys
+    """
+    if isinstance(docs, list):
+        results = []
+        for doc in docs:
+            if doc:
+                for key in stripKeys:
+                    doc.pop(key, None)
+                    results.append(doc)
+        return results
+
+    if docs and isinstance(docs, dict):
+        for key in stripKeys:
+            docs.pop(key, None)
+        return docs
+
+class MSPileupData(MSCore):
+    """
+    MSPileupData class provide whole logic behind data used and stored by MSPileup module
+    """
+
+    def __init__(self, msConfig, **kwargs):
+        super(MSPileupData, self).__init__(msConfig, **kwargs)
+
+        self.msConfig.setdefault("mongoDBRetryCount", 3)
+        self.msConfig.setdefault("mongoDBReplicaSet", None)
+        self.msConfig.setdefault("mongoDBPort", None)
+        self.msConfig.setdefault("mockMongoDB", False)
+
+        # A full set of valid database connection parameters can be found at:
+        # https://pymongo.readthedocs.io/en/stable/api/pymongo/mongo_client.html
+        mongoIndex = IndexModel('pileupName', unique=True)
+        collection = self.msConfig.get('mongoDBCollection', 'msPileupCollection')
+        msOutDBConfig = {
+            'database': self.msConfig['mongoDB'],
+            'server': self.msConfig['mongoDBServer'],
+            'replicaSet': self.msConfig['mongoDBReplicaSet'],
+            'port': self.msConfig['mongoDBPort'],
+            'username': self.msConfig['mongoDBUser'],
+            'password': self.msConfig['mongoDBPassword'],
+            'mockMongoDB': self.msConfig['mockMongoDB'],
+            'connect': True,
+            'directConnection': False,
+            'logger': self.logger,
+            'create': True,
+            'collections': [(collection, mongoIndex)]
+        }
+        mongoDB = MongoDB(**msOutDBConfig)
+        self.msDB = getattr(mongoDB, self.msConfig['mongoDB'])
+        self.dbColl = self.msDB[collection]
+#         self.dbColl.create_indexes([mongoIndex])
+
+    def create(self, cdict):
+        """
+        Create and return pileup data from campaigns dictionary
```

Comment:
We'd better think of continuous parsing of the campaign config and updating of all relevant documents in the database based on the information obtained in the process and reflecting also the current  load in the system. So again, this needs to go as a step in the service.

Reply:
No, service can accept campaigns.json, parse it and only call data layer to insert docs, e.g.
```
class MSPileupService():
    def __init__(self):
           # it keeps internal access to DB but it is re-used in this class across APIs, i.e. single session
           self.mspileup = MSPileupData()
     def campaignAPI(self, campaigns):
          # parse campaigns:
          for doc in parse(campaigns:)
               doc = self.mspileup.createPileupObj(doc)
               self.mspileup.strorePileupObj(doc)
```
Nice and clear separation of layers, the service accept campaigns, and parse it. Then it calls data layer to insert docs.

Reply:
One note here, though. I wouldn't  call this method  `campaignAPI`, because it should not be supposed to provide inter-service communication API (I can imagine only internal service calls for this one). Maybe calling it  `campaingParser` would be more direct and pointing to it's exact and sole purpose. Just an idea...